Головнев Максим

maximka200 opened the pull request:
Планирую такую логику:
 - Текст на вход приходит в Lexer, где он переводится в List<Token> (список слов, пробелов, значащих символов)
 - List<Token> приходит в Parser, откуда приходит уже выровненный по структуре и иерархии корень дерева типа Node (дерево позволит нам определять вложенные теги)
 - Затем из этого дерева генерится html в HtmlGenerator

maximka200 committed:
init commit: накидал начальную структуру.
Планирую такую логику:
 - Текст на вход приходит в Lexer, где он переводится в List<Token> (список слов, пробелов, значащих символов)
 - List<Token> приходит в Parser, откуда приходит уже выровненный по структуре и иерархии корень дерева типа Node (дерево позволит нам определять вложенные теги)
 - Затем из этого дерева генерится html в HtmlGenerator

maximka200 committed:
- Сделал все классы (кроме тех, что лежат в Domains) статическими
- Добавил проект с тестами и написал простые тесты на MdLexer.cs
- Реализовал логику считывания токенов в MdLexer.cs (слова, #, _)

maximka200 committed:
Добавил тестов и вынес метод CollectFullValue в MdLexer.cs

Dimques commented on cs/Markdown/Lexer/MdLexer.cs:
```diff
@@ -0,0 +1,63 @@
+                        tokenType is TokenType.Word ? char.IsLetter : char.IsNumber
+                        );
+                    tokens.Add(new MdToken(tokenType, value));
+                    i = nextIndex - 1;
```
Давай вот эту работу с вычитанием единицы спрячем в `CollectFullValue`.

Dimques commented on cs/Markdown/Lexer/MdLexer.cs:
```diff
@@ -0,0 +1,63 @@
+
+    private static (string word, int nextIndex) CollectFullValue(string text, int startIndex, Func<char, bool> predicate)
+    {
+        var word = new StringBuilder();
```
Ну это же не обязательно word, оставь value лучше.

Dimques commented on cs/Markdown/Lexer/MdLexer.cs:
```diff
@@ -0,0 +1,63 @@
+        return (word.ToString(), i);
+    }
+
+    public static TokenType GetTokenType(char text) =>
```
Не забудь, что помимо Space есть и другие пробельные символы.

Dimques commented on cs/Markdown/Lexer/MdLexer.cs:
```diff
@@ -0,0 +1,63 @@
+            '#' => TokenType.Grid,
+            '_' => TokenType.Underscore,
+            ' ' => TokenType.Space,
+            _ => throw new ArgumentException($"Unknown token type: {text}")
```
Дело вкуса, но можно и `ArgumentOutOfRangeException` выкидывать.

Dimques commented on cs/Markdown/Markdown.cs:
```diff
@@ -0,0 +1,17 @@
+
+public class Markdown
+{
+    public string ConvertToHtml(string text)
```
Читаю задание: "В fork-е этого репозитория создай проект Markdown и реализуй метод Render класса Md". То, что ты назвал произвольным образом не грубая ошибка, но затрудняет проверку.

Dimques commented on cs/MarkdownTests/LexerTests.cs:
```diff
@@ -0,0 +1,83 @@
+            "#Test",
+            new List<MdToken>
+            {
+                new MdToken(TokenType.Grid, "#"),
```
Тут компилятор сам может вывести название создаваемого типа, можно просто new(TokenType.Grid, "#").

Dimques commented on cs/MarkdownTests/LexerTests.cs:
```diff
@@ -0,0 +1,83 @@
+                new MdToken(TokenType.Underscore, "_"),
+                new MdToken(TokenType.Word, "Test")
+            }
+        ).SetName("WithUnderscoreAndWord_1");
```
`WithGridAndWord_1` - а что значит 1?

maximka200 replied:
хотел сделать еще тесткейсов

Dimques replied:
Тогда возможно будет лучше вынести такие кейсы в отдельный сорс и сделать отдельный тест с названием про `UnderscoreAndWord`

Dimques commented on cs/MarkdownTests/LexerTests.cs:
```diff
@@ -0,0 +1,83 @@
+    
+    [Test]
+    [TestCaseSource(nameof(LexerTokenizeTestCases))]
+    public void Tokenize_ActualOutputShouldBeEqualExpectedOutput(string input, List<MdToken> expectedOutput)
```
И так понятно, что в тесте какой-то актуал должен соответствовать экспектеду. Если ничего на ум не идёт называй просто а-ля Lexer_should_tokenize_correctly.

maximka200 committed:
- Создал класс HeaderNode - наследника Node
- Создал класс TokenParserTests.cs

maximka200 committed:
правки замечаний

maximka200 committed:
- Вынес TextNode в отдельный класс
- Дописал типы Node в NodeType
- Дописал типы Token в TokenType
- Убрал поле text из базовой ноды

maximka200 committed:
- fix TextNode.cs

maximka200 committed:
- Написал тесты
- Написал базовую реализацию в декларативном стиле и без некоторой логики, поэтому не все тесты проходят

maximka200 committed:
- переписал некоторые тесты
- добавил класс-раширение для List<MdToken>
- более менее декомпозировал метод Parse
- не работает условие с одинарными тегами внутри двойных/наоборот

maximka200 committed:
сделал корректное определение подчеркивания в словах с числами, разных словах

maximka200 committed:
- написал еще методов в ListMdTokenExtension.cs и поправил старые
- убрал некорректный тест в ListMdTokenTests.cs
- сделал логику чтобы работало условие на вложенность Bold в Italic/наоборот
- поправил тесты в TokenParserTests.cs

maximka200 committed:
косметические правки

maximka200 committed:
- написал тест на HtmlGenerator.cs
- написал HtmlGenerator.cs

maximka200 committed:
Декомпозировал Parse

maximka200 committed:
- Написал сквозные тесты
- Пофиксил некорректное экранирование
- Косметические правки

maximka200 committed:
написал перформанс тест

maximka200 committed:
- реализовал ссылки (еще не написал по ним спеки)
- дописал перфоманс тест

maximka200 committed:
- улучшил читаемость
- убрал ref из GetLengthChainOfTokenTypeAfter

maximka200 committed:
написал спеки для гиперссылки

Dimques commented on cs/Markdown/Domains/NodeExtensions/HeaderNode.cs:
```diff
@@ -0,0 +1,20 @@
+namespace Markdown.Domains.NodeExtensions;
```
Слово Extensions обычно в другом смысле используют - для классов с методами-расширениями. Давай переименуем как-нибудь более соответствующе директории-хранилищу наследников ноды.

maximka200 replied:
например, NodeTypes?

Dimques replied:
Тогда будет немного непонятно, почему файлы `LinkNodeType` и `NodeType` снаружи лежат. Но да, уже получше.

Dimques commented on cs/Markdown/Domains/TokenType.cs:
```diff
@@ -0,0 +1,17 @@
+    Asterisk,
+    Escape,
+    Slash,
+    LeftSquareBracket, // [
```
Лишний комментарий. Вообще, как правило, комментарии нужны не чтобы объяснить, что делает код (хотя и такое бывает), а чтобы объяснить, по каким причинам он делает именно так, а не иначе.

maximka200 replied:
ну я больше для себя написал, т.к. переодически путал left и right))

Dimques commented on cs/Markdown/Generator/HtmlGenerator.cs:
```diff
@@ -0,0 +1,48 @@
+    {
+        return node.Type switch
+        {
+            NodeType.Root => string.Concat(node.Children.Select(Generate)),
```
1. Ну т.е. ответственность генератора - выяснить тип каждой ноды и запустить обход детей? А, ещё он знает, как текст каждой ноды надо декорировать перед обходом детей. Начну чуть издалека - скажи, ты слышал что-нибудь про полиморфизм?
2. Спускаюсь по стек-трейсу - string.Concat(string.Concat(string.Concat(string.Concat(...)))). Не многовато ли памяти выделится?

maximka200 replied:
1. Понял, наверное добавлю в каждый класс ноды свой ToHtml и буду его в генераторе вызывать
2. Тут поменяю string -> StringBuilder

Dimques commented on cs/Markdown/Lexer/MdLexer.cs:
```diff
@@ -0,0 +1,78 @@
+            '#' => TokenType.Grid,
+            '*' => TokenType.Asterisk,
+            '_' => TokenType.Underscore,
+            ' ' or '\u00a0' or '\u200b' => TokenType.Space,
```
Интересный ход мысли. Если что - я на табуляцию намекал :)

Dimques commented on cs/Markdown/Lexer/MdLexer.cs:
```diff
@@ -0,0 +1,78 @@
+
+    private static bool IsLetterOrDotСomma(this char ch)
+    {
+        return char.IsLetter(ch) || ch == '.' || ch == ',';
```
Может, другие знаки препинания тоже обработаем?

Dimques commented on cs/Markdown/Parser/ListMdTokenExtension.cs:
```diff
@@ -0,0 +1,123 @@
+    ///     Поиск закрывающего индекса закрывающего тега
+    /// </summary>
+    /// <param name="startIndex">Индекс первого токена после конца открывающей цепочки токенов</param>
+    /// <returns>Индекс первого токена в закрывающей цепочке</returns>
```
А почему ты только один параметр описал?
Давай еще опишем где-то, что означает возвращаемое значение -1 и почему метод может выбросить ошибку.

Dimques commented on cs/Markdown/Parser/ListMdTokenExtension.cs:
```diff
@@ -0,0 +1,123 @@
+public static class ListMdTokenExtension
+{
+    /// <summary>
+    ///     Поиск закрывающего индекса закрывающего тега
```
Дак тут про тэги или про токены? :)

Dimques commented on cs/Markdown/Parser/ListMdTokenExtension.cs:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    /// <summary>
+    ///     Проверка на то, что под__черкивания в раз__ных словах, _разных сл_овах, раз_ных словах_
```
Название метода и так довольно говорящее. Давай лучше опишем, что ты передаёшь в качестве аргументов, что возвращаешь и когда метод может выкидывать ошибку - по аналогии с предыдущим методом.

Dimques commented on cs/Markdown/Parser/ListMdTokenExtension.cs:
```diff
@@ -0,0 +1,123 @@
+        return hasSeparator && (prevIsWord || nextIsWord);
+    }
+
+    public static bool IsUnderscoreInWordWithNumbers(this List<MdToken> tokens, int startIndex, int closeIndex,
```
Ну раз уж метод публичный, тогда и ему xml-doc давай добавим.

Dimques commented on cs/Markdown/Parser/ListMdTokenExtension.cs:
```diff
@@ -0,0 +1,123 @@
+
+    internal static void AddSymbol(this List<Node> root, string symbol, int count)
+    {
+        for (var _ = 0; _ < count; _++)
```
Даже неиспользуемый `i` в цикле намного привычнее взгляду, чем `_`.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+    public static Node Parse(List<MdToken> tokens, NodeContext context = NodeContext.None)
+    {
+        var rootChildren = new List<Node>();
+        var i = 0;
```
А чего ты этому `i` осмысленное имя не дашь?

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+        return new Node(NodeType.Root, rootChildren);
+    }
+
+    private static void HandleHeader(List<MdToken> tokens, ref int i, List<Node> rootChildren)
```
Довольно мучительно получается прокидывать этот `int i` по цепочке вызовов, да и список тоже. Давай сделаем следующее:
1. класс сделаем нестатическим
2. сделаем `i` приватной переменной
3. добавим метод `InitializeWith` (или что-то в таком духе), который будет принимать токены и сохранять их в приватное поле, а так же сбрасывать позицию `i` в 0
4. добавим метод, который будет увеличивать позицию и заиспользуем его везде, где мы сейчас увеличиваем позицию через те или иные операции
5. придумаем методы, чтобы заменить штуки, вроде `i + 1` (это же какая-то позиция по отношению к текущей, да?)

Dimques commented on cs/Markdown/Parser/ListMdTokenExtension.cs:
```diff
@@ -0,0 +1,123 @@
+
+        var range = tokens.GetRange(startIndex + underscoreCount, closeIndex - startIndex);
+
+        var hasNumber = range.ContainsTokenType(TokenType.Number);
```
1. давай уберём пустые строки между переменными
2. переменную `range` можно выкинуть, написать сразу `var hasNumber = tokens.GetRange(startIndex + underscoreCount, closeIndex - startIndex).ContainsTokenType(TokenType.Number);`

Dimques commented on cs/Markdown/Parser/ListMdTokenExtension.cs:
```diff
@@ -0,0 +1,123 @@
+            root.Add(new TextNode(symbol));
+    }
+
+    internal static int GetLengthChainOfTokenTypeAfter(this List<MdToken> tokens, int startIndex, TokenType tokenType)
```
`GetTokensCountAfter`

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+        if (spaceCountAfter != 0)
+        {
+            rootChildren.AddSymbol("_", underscoreCount);
+            rootChildren.AddSymbol(" ", underscoreCount);
```
После подчёркиваний добавляешь пробелов столько же, сколько было подчёркиваний. Баг?

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+            return;
+        }
+
+        switch (underscoreCount)
```
Давай заменим `switch` на `if (underscoreCount is 1 or 2)`

Dimques commented on cs/Markdown/Parser/ListMdTokenExtension.cs:
```diff
@@ -0,0 +1,123 @@
+        return hasNumber && prevIsWord;
+    }
+    
+    internal static int GetLengthChainOfTokenTypesBefore(this List<MdToken> tokens, int index, TokenType tokenType)
```
`GetTokensCountBefore`

Dimques commented on cs/Markdown/Parser/ListMdTokenExtension.cs:
```diff
@@ -0,0 +1,123 @@
+                }
+
+            if (!match)
+                continue;
```
Давай вынесем этот код в отдельный метод `CheckMatch`, который будет возвращать булево значение, т.е. вместо всего этого будет `if (!CheckMatch) continue;`

Dimques commented on cs/Markdown/Parser/ListMdTokenExtension.cs:
```diff
@@ -0,0 +1,123 @@
+            var prevIsSame = j - 1 >= 0 && tokens[j - 1].Type == tokenType;
+            var nextIsSame = j + patternLen < tokens.Count && tokens[j + patternLen].Type == tokenType;
+
+            if (prevIsSame || nextIsSame)
```
Смотри, можно написать `if (!prevIsSame && !nextIsSame) return j;` и надобность в `continue` отпадёт. Но двойное отрицание через && в if тяжело читается. Может быть можно назвать переменные по-другому и немного другие проверки сделать, чтобы избавиться от лишнего `continue`?

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+            rootChildren.AddSymbol("[", 1);
+            i++;
+            return;
+        }
```
Этот код - отдельный логический шаг, придумай, как его вынести в метод.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+        i++;
+    }
+
+    private static void HandleLeftSquareBracket(List<MdToken> tokens, ref int i, List<Node> rootChildren,
```
Очень громоздкий получился метод.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+        var nextIsLeftParen = closeIndexMeaningText < tokens.Count - 1 &&
+                              tokens[closeIndexMeaningText + 1].Type == TokenType.LeftParenthesis;
+
+        if (nextIsLeftParen && closeIndexMeaningText + 2 < tokens.Count)
```
Переменную `nextIsLeftParen` выкинуть, а то, из чего она состоит и вторую часть из `if`-а - в отдельный булевый метод выноси, т.е. пусть этот метод содержит в себе всё, что в `if`-е проверяется сейчас. Придумай хорошее название, чтобы было понятно, что проверяется, но не супер-длинно. `+2` здесь это какая-то магия, на первый взгляд - возможно, ты одним словом в названии проверки объяснишь, что это за двойка, подумай.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+    private static void HandleLeftSquareBracket(List<MdToken> tokens, ref int i, List<Node> rootChildren,
+        NodeContext context)
+    {
+        var squareBracketsLength = tokens.GetLengthChainOfTokenTypeAfter(i, TokenType.LeftSquareBracket);
```
По аргументу и так понятно, что это квадратная скобка, пусть переменная будет просто `bracketsLength`.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+                    new LinkNode(LinkNodeType.MeaningText, meaningText),
+                    new LinkNode(LinkNodeType.LinkText, linkText),
+                });
+
```
Этот код - отдельный логический шаг, давай вынесем его в метод `BuildLinkNode`.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+
+                rootChildren.Add(linkNode);
+                i = closeIndexLinkText + 1;
+                return;
```
Кажется, можно убрать `return` и добавить  `else` к нужному `if`-у?

maximka200 replied:
тут тогда не очень красиво получится, что то вроде:
```
if (...)
{
    if (...)
    {
        ...
    {
     else
    {
        rootChildren.AddSymbol("[", 1);
        i++;
    }
}
else
{
   rootChildren.AddSymbol("[", 1);
   i++;
}
```

> Кажется, можно убрать `return` и добавить `else` к нужному `if`-у?

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+
+
+    private static int HandleNonFormattingUnderscore(
+        List<MdToken> tokens,
```
Не, давай в одном стиле сделаем, пусть в строчку будут аргументы.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+    {
+        var closeIndex = tokens.FindClosing(i, underscoreCount, TokenType.Underscore);
+
+        if (closeIndex != -1)
```
Давай наоборот сделаем - `if (closeIndex == 1)` то добавь `underscore`-ы и сделай `return`, а уже после давай поставим блок, который сейчас внутри `if`-а - так ты вложенность снизишь, т.е. у тебя огромный кусок кода в `if`-е, это плохо читается, потому что тебе как-бы надо держать в голове помимо самого большого куска тот факт, что он выполняется только в случае успешной проверки. А если инвертнуть `if`, как я предлагаю, то читателю сразу ясно - вот мы вначале проверили какой-то `edge-кейс` и уже после этого идём дальше спокойно.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+            // проверка на пробелы до
+            if (spaceCountBefore > 0)
+                return HandleNonFormattingUnderscore(tokens, i, closeIndex, underscoreCount, spaceCountBefore,
+                    rootChildren);
```
Оставь в одной строке - лучше пусть она чуть длинновата будет, чем образует ступеньку из одного аргумента.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+            var nodeType = underscoreCount == 1 ? NodeType.Italic : NodeType.Bold;
+
+            // проверка на _ в разных словах, _ в словах с числами
+            if (context == NodeContext.Italic
```
Проверку в `if`-е  вынеси в метод с названием примерно соответствующим комментарию, а коммент убери.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+                return HandleNonFormattingUnderscore(tokens, i, closeIndex, underscoreCount, spaceCountBefore,
+                    rootChildren);
+
+            var nodeType = underscoreCount == 1 ? NodeType.Italic : NodeType.Bold;
```
Объяви прямо перед переменной `inner`.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,251 @@
+            var inner = Parse(tokens.GetRange(i, closeIndex - i), Node.GetNodeContext(nodeType));
+
+            rootChildren.Add(new Node(nodeType, inner.Children));
+            return closeIndex - i + underscoreCount;
```
1. вместе с `nodeType` этот код образует отдельный логический шаг
2. непонятно, что за `range` ты передаёшь в `parse` здесь - вынеси его в отдельную переменную с названием, которое поможет это понять
3. что за `inner`?
4.  `closeIndex - i + underscoreCount` - давай для этого возвращаемого значения придумаем понятное название переменной, а то я пока читал метод, забыл, что он возвращает

Dimques commented on cs/MarkdownTests/Lexer/LexerTests.cs:
```diff
@@ -0,0 +1,106 @@
+    [TestCase('#', TokenType.Grid)]
+    [TestCase('_', TokenType.Underscore)]
+    [TestCase('1', TokenType.Number)]
+    public void GetTokenType_ShouldGetToken_Correctly(char input, TokenType expectedOutput)
```
1. Мало кейсов
2. Нужен отдельный тест с кейсами, когда `GetTokenType` выкинет  ошибку

Dimques commented on cs/MarkdownTests/MarkdownEndToEndTests.cs:
```diff
@@ -0,0 +1,25 @@
+namespace MarkdownTest;
+
+[TestFixture]
+public class MarkdownEndToEndTests
```
Не критично, но всё таки я бы сказал, что это скорее функциональные тесты, чем end2end. Ты в функциональном тесте ты проверяешь работу нескольких связанных модулей, в end2end - приложение целиком. Вот если бы ты в тесте через `shell` запускал программу, которой на вход передавал бы, например, аргументы с именем файла входных дайнных, именем файла выходных данных и получившийся выходной файл после сравнил с каким-то файлом-образцом - вот это был бы end2end.

Dimques commented on cs/MarkdownTests/MarkdownEndToEndTests.cs:
```diff
@@ -0,0 +1,25 @@
+    [TestCase("[text](example.com)", "<a href=\"example.com\">text</a>")]
+    [TestCase("Before [text](example.com) after", "Before <a href=\"example.com\">text</a> after")]
+    [TestCase("Nested __[bold](link.com)__", "Nested <strong><a href=\"link.com\">bold</a></strong>")]
+    public void Should_ConvertMarkdownToHtml_Correctly(string markdown, string expectedHtml)
```
1. А если пробелов между `#` и тесктом будет больше одного? Маркдаун должен ведь все пробельные символы скушать, нет?
2. А если между `#` и текстом будет табуляция?
3. А если я эскейп поставлю перед гридом и уберу после него пробел? :)
4. Уровней заголовков по коду 6, здесь только 3 проверяется - давай отдельный тест сделаем под это. Кстати, почему на 7 заголовков выкидывается ошибка? Я ведь тогда, зная синтаксис, не смогу начать строку с произвольного числа grid-ов - а что, если я часть текста выделяю строками с десятками "#####...#"?
5. Не работает кейс, где экранируется закрывающий тэг, т.е. например `[TestCase(@"Escaped _underscore\_", "Escaped _underscore_")]`. И похоже, что это не зависит от типа символа.
6. `[TestCase("root1_2_3", "root1_2_3")]` - вот так не сработает. То же самое про стронг.
7. Повторюсь - очень неожиданно, что на восклицательные и вопросительные знаки здесь вылетают эксепшоны.
8. Судя по спеке ("Внутри текста ссылки и адреса ссылки могут использоваться любые символы разметки, которые будут обработаны по общим правилам:"), у меня должны работать вложенные скобки - т.е. скобки это же символы разметки, значит внутри других скобок они будут обработаны по общим правилам, в которых ничего про скобки нет. Либо давай уточним спеку.

Мало примеров с новым тэгом и хитрыми пересечениями. И еще - давай разобьем по группам всё-таки тесты, а то немного в глазах рябит сейчас. А если подряд будут только текст-кейсы, например, со скобками, то будет проще гораздо оценить корректность теста.

Dimques commented on cs/MarkdownTests/HtmlGenerator/HtmlGeneratorTests.cs:
```diff
@@ -0,0 +1,43 @@
+    [TestCaseSource(nameof(HtmlGeneratorTestCases))]
+    public void Parse_ShouldParse_Correctly(Node node, string expectedText)
+    {
+        var result = Markdown.Generator.HtmlGenerator.Generate(node);
```
Придумай, как сделать так, чтобы не надо было писать в тесте неймспейс.

Dimques replied:
Ну и?

maximka200 replied:
в другом файл пофиксил,
тут что-то забыл :)

> Ну и?

Dimques commented on cs/MarkdownTests/MarkdownPerformanceTests.cs:
```diff
@@ -0,0 +1,48 @@
+public class MarkdownPerformanceTests
+{
+    [Test]
+    public void Markdown_Render_ShouldWorkFast()
```
Тест флакает, т.е. нестабильно работает - добавь `[Repeat(100)]` и скорее всего заметишь. Вообще, для тестов со случайным наобором данных так принято делать, хоть это и не 100% гарантия. Но вообще ты здесь можешь придумать просто какой-нибудь худший пример - асимптотика ведь по нему оценивается. А еще, даже если у тебя честная асимптотика nlogn, то реальный пример может не укладываться в неё, потому что асимптотика не учитывает константу, которая в реальных примерах всегда есть.

Dimques commented on cs/MarkdownTests/MarkdownPerformanceTests.cs:
```diff
@@ -0,0 +1,48 @@
+            var markdown = GenerateRandomMarkdown(length);
+            sw.Start();
+
+            Markdown.Markdown.Render(markdown);
```
Придумай, как сделать так, чтобы не использовать в тестах полное имя типа.

Dimques commented on cs/MarkdownTests/MarkdownPerformanceTests.cs:
```diff
@@ -0,0 +1,48 @@
+    private static string GenerateRandomMarkdown(int len)
+    {
+        var rand = new Random();
+        var specElements = new[] { " ", "_", "__", "#", "\\", "\\n", "(", ")", "[", "]" };
```
Чтобы ничего не забыть, можно составить массив из символов по ascii-кодам, только границы кодов надо правильно выбрать.

Dimques commented on cs/MarkdownTests/MarkdownPerformanceTests.cs:
```diff
@@ -0,0 +1,48 @@
+
+    private static string GenerateRandomMarkdown(int len)
+    {
+        var rand = new Random();
```
Не критично, но посмотри про разницу между `new Random()` и `Random.Shared`.

maximka200 committed:
- правки по MdLexer.cs
- правки по ListMdTokenExtension.cs

maximka200 committed:
- правки по TokenParser.cs

maximka200 committed:
- уточнил спеки
- переписал перформанс тест, теперь он не флокает
- разбил функциональные тесты
- пофиксил баг в IsUnderscoreInWordWithNumbers

maximka200 committed:
переименовал неймспейс NodeTypes -> Nodes

maximka200 committed:
дописал тестов на header и пофиксил методы для его парсинга

Dimques commented on cs/Markdown/Domains/IToHtml.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace Markdown.Domains;
+
+public interface IToHtml
```
Очень скупое название. Почему не назвать интерфейс просто `IHtmlConverter` и не сделать ему метод `ConvertToHtml`?

Dimques commented on cs/Markdown/Domains/Node.cs:
```diff
@@ -0,0 +1,70 @@
+    private NodeType Type { get; } = type;
+    public List<Node> Children { get; } = children ?? [];
+
+    public virtual void ToHtml(StringBuilder sb)
```
Направление правильное, но это не совсем то, что я имел ввиду. Такой свитч у тебя и так под капотом (т.е. после компиляции) будет, если вместо `NodeType` ты просто сделаешь ноды таких же типов, как сейчас в `enum`, унаследованных от `Node`, например `RootNode : Node` и у каждой переопределишь `ToHtml` в соответствии с тем, что в этом свиче написано. Т.е. под полиморфизмом имелось ввиду, что есть ноды, у которых есть дети-ноды и что любая нода не знает, какого именно типа её дети, и ей и не надо этого знать, она просто знает, что все они конвертятся в html (ведь ей доступен этот метод) и вызывает его, а как уж он там реализован - это знает только сам нода-ребёнок и будет вести себя соответственно - например, если это bold-нода, то конвертирование будет работать с тэгом `strong` и т.п.

Dimques replied:
И этот механизм в язык внедрён как раз, чтобы не делать так, как сделал ты, т.е. не плодить дополнительные сущности-`enum`-ы и свичи по ним. Да, иногда без этого никак и всё равно приходится писать свич по типу класса, но здесь можно было обойтись без него.

Dimques commented on cs/Markdown/Domains/Node.cs:
```diff
@@ -0,0 +1,70 @@
+        };
+    }
+
+    private void RenderRoot(StringBuilder sb)
```
Ну вот, опять же - пришлось пачку разных рендер-методов делать в одной ноде. А так бы каждая реализация была в своём классе и название было бы одно общее `Render` и может даже интерфейс был бы не нужен.

Dimques commented on cs/Markdown/Domains/TokenType.cs:
```diff
@@ -0,0 +1,18 @@
+    Asterisk,
+    Escape,
+    Slash,
+    Tab,
```
Выделять `Tab` в отдельный тип - костыль. А если я ещё пробельных символов придумаю? Может, в каком-то другом месте всё-таки должна быть проверка, является ли символ пробельным и по результатам этой проверки одинаковое поведение для любого пробельного символа?

Dimques commented on cs/Markdown/Generator/HtmlGenerator.cs:
```diff
@@ -0,0 +1,17 @@
+    public static string Generate(IToHtml node)
+    {
+        var sb = new StringBuilder();
+        node.ToHtml(sb);
```
Что-то мне подсказывает, что ответственность этого класса - надуманная и можно её перенести на `RootNode`, не создавая снаружи стрингбилдор, а генератор этот выкинуть к чёртовой бабушке.

Dimques commented on cs/Markdown/Lexer/MdLexer.cs:
```diff
@@ -0,0 +1,84 @@
+        { '_', TokenType.Underscore },
+        { ' ', TokenType.Space },
+        { '\u00a0', TokenType.Space },
+        { '\u200b', TokenType.Space },
```
Ну вот как так - другие пробельные символы у тебя спейс, а таб это таб??

Dimques commented on cs/Markdown/Lexer/MdLexer.cs:
```diff
@@ -0,0 +1,84 @@
+    private static readonly Dictionary<char, TokenType> TokenMap = new()
+    {
+        { '#', TokenType.Grid },
+        { '*', TokenType.Asterisk },
```
Зачем звёздочке отдельный тип?

Dimques commented on cs/Markdown/Lexer/MdLexer.cs:
```diff
@@ -0,0 +1,84 @@
+namespace Markdown.Lexer;
+
+/// <summary>
+///     Разбивает текст на Md токены
```
Было бы неплохо здесь эти токены перечислить

Dimques commented on cs/Markdown/Markdown.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace Markdown;
+
+public static class Markdown
```
Вот я с самого начала тебе указывал - по заданию надо, чтобы класс `Md` был. У тебя теперь один класс `Markdown`, а второй `MdLexer`🗿 Поверь, если бы было 100500 файлов в проекте, то никто бы не понял, что эти классы имеют связь какую-то.

Dimques commented on cs/Markdown/Parser/ListMdTokenExtension.cs:
```diff
@@ -0,0 +1,149 @@
+
+public static class ListMdTokenExtension
+{
+    /// <summary>
```
Ну вот, другое дело 👍 
<img width="671" height="657" alt="image" src="https://github.com/user-attachments/assets/27959559-5835-4052-90f5-85a3bb3211c6" />

Dimques commented on cs/Markdown/Parser/ListMdTokenExtension.cs:
```diff
@@ -0,0 +1,149 @@
+        return hasNumber && prevIsWord;
+    }
+
+    internal static int GetTokensCountBefore(this List<MdToken> tokens, int index, TokenType tokenType)
```
Давай `internal`-модификатор осознанно использовать - здесь непонятно, почему часть методов `internal`, часть `private` и ещё часть `public`. Убедись, что понимаешь разницу между `internal` и `private`.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,316 @@
+namespace Markdown.Parser;
+
+/// <summary>
+///     Парсит лист токенов в синтаксическое дерево
```
`лист` - ну ладно хоть не в `синтакс три` парсит. Пойми меня правильно - в англицизмах ничего плохого нет, но у слова `list` есть адекватный, широкораспространённый перевод, это слово `список`.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,316 @@
+        rootChildren.Add(new HeaderNode(level, innerBlock));
+    }
+
+    private void Move(int shift = 1)
```
`Move` куда? Назови так, чтобы без чтения тела метода можно было понять, что он делает.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,316 @@
+        Move();
+    }
+
+    private void HandleLeftSquareBracket(List<Node> rootChildren, NodeContext context)
```
Семантика `Try`-методов подразумевает, что они возвращают булево значение и имеют `out`-параметр, который равен `null` если метод возвращает `false`, и имеет актуальное значение, если метод возвращает `true`. Здесь это могло бы выглядеть, например, вот так:
```
    private void HandleLeftSquareBracket(List<Node> rootChildren, NodeContext context)
    {
        if (TryParseLink(context, out var linkNode))
        {
            rootChildren.Add(linkNode!);
        }
        else
        {
            rootChildren.AddSymbol("[", 1);
            Move();
        }
    }
``` 
....
```
    private bool TryParseLink(NodeContext context, out LinkNode? node)
    {
        node = null;
        
        var bracketsLength = tokens.GetTokensCountAfter(position, TokenType.LeftSquareBracket);
        var closeIndexMeaningText = tokens.FindClosing(position, bracketsLength, TokenType.RightSquareBracket);
        
        if (closeIndexMeaningText == -1 || IsTokenType(closeIndexMeaningText - 1, TokenType.Slash)
            || !IsValidLinkSyntax(closeIndexMeaningText))
            return false;

        var closeIndexLinkText = tokens.FindClosing(closeIndexMeaningText + 2, 1, TokenType.RightParenthesis);
        if (closeIndexLinkText == -1 || IsTokenType(closeIndexLinkText - 1, TokenType.Slash))
            return false;

        node = BuildLinkNode(
            meaningStart: position + 1,
            meaningEnd: closeIndexMeaningText,
            linkStart: closeIndexMeaningText + 2,
            linkEnd: closeIndexLinkText,
            context: context
        );

        Move(closeIndexLinkText + 1 - position);

        return true;
    }
```

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,316 @@
+            || !IsValidLinkSyntax(closeIndexMeaningText))
+            return null;
+
+        var closeIndexLinkText = tokens.FindClosing(closeIndexMeaningText + 2, 1, TokenType.RightParenthesis);
```
Заметил, что ты в обратном порядке английские слова постоянно расставляешь - это, как ни странно, снижает читаемость. Главное существительное должно в самом конце стоять, а перед ним по цепочке существительные и прилагательные, которые имеют принадлежность к нему (сорри, если криво формулирую - не знаток лингвистики), т.е. если ты хочешь сказать "закрывающий индекс текста ссылки", то надо формулировать `LinkTextCloseIndex`. А тут я читаю `ссылочный текст закрывающего индекса`. В таком порядке можно оставить, если ты `Of` используешь, т.е. тогда ты как бы указываешь, что индекс это главное слово: `closeIndexOfLinkText`. Но это удлинняет наименование, поэтому обычно выстраивают слова в обратном порядке. Кстати, попарно они у тебя правильно поставлены - `LinkText` это текст ссылки, а `CloseIndex` это индекс закрытия или закрывающий индекс.

Dimques replied:
Ещё пояснение, почему такое именование снижает читаемость - вижу `closeIndexMeaningText - 1`, автоматом считываю, что `из какого-то текста вычитается единица`.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,316 @@
+        return new Node(NodeType.Root, rootChildren);
+    }
+    
+    private static TokenParser InitializeWith(List<MdToken> inputTokens)
```
В итоге декоратор конструктора получился. Моя идея была в том, чтобы перенести куда-то инициализацию позиции из метода `Parse`, чтобы его рекурсивно можно было использовать, а так же чтобы был метод, который как-бы рестартовал работу уже созданного экземпляра `TokenParser`, чтобы не надо было его пересоздавать. Ну ладно.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,316 @@
+
+    private void HandleSlashCharacter(List<Node> rootChildren)
+    {
+        if (position + 1 < tokens.Count)
```
Я ведь предлагал для `position + 1` сделать методы. Тут хотя бы локальная переменная нужна, чтобы если вдруг придётся менять значение (например, на `position + 2`) не было шансов забыть во всех местах поменять.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,316 @@
+            return null;
+
+        var closeIndexLinkText = tokens.FindClosing(closeIndexMeaningText + 2, 1, TokenType.RightParenthesis);
+        if (closeIndexLinkText == -1 || IsTokenType(tokens, closeIndexLinkText - 1, TokenType.Slash))
```
Кажется, незачем в `IsTokenType` явно токены передавать. Ниже подробнее напишу.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,316 @@
+        if (closeIndexLinkText == -1 || IsTokenType(tokens, closeIndexLinkText - 1, TokenType.Slash))
+            return null;
+
+        var linkNode = BuildLinkNode(
```
Использовал имена аргументов при инициализации, это лайк👍

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,316 @@
+    private int HandleNonFormattingUnderscore(int closeIndex, int underscoreCount, int spaceCountBefore,
+        List<Node> rootChildren, NodeContext context = NodeContext.None)
+    {
+        var tokensInside = tokens.GetRange(position, closeIndex - position - spaceCountBefore);
```
Токены внутри чего?

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,316 @@
+        List<Node> rootChildren, NodeContext context = NodeContext.None)
+    {
+        var tokensInside = tokens.GetRange(position, closeIndex - position - spaceCountBefore);
+        var innerBlock = InitializeWith(tokensInside)
```
Вот здесь против тебя играет, что неясно, что `InitializeWith` новый экземпляр парсера создаёт. Давай его переименуем тогда в `CreateFor` (и в скобках как бы имя переменной, т.е. для чего создаётся - для вот этих вот токенов) или просто в `CreateParser` .

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,316 @@
+    {
+        var tokensInside = tokens.GetRange(position, closeIndex - position - spaceCountBefore);
+        var innerBlock = InitializeWith(tokensInside)
+            .Parse(context).Children;
```
В одну строчку норм смотрится, ниже ты так же делаешь.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,316 @@
+        return count;
+    }
+    
+    private static bool IsTokenType(List<MdToken> tokenList, int index, TokenType type)
```
Сделай оба метода нестатическими, тогда не надо будет список в аргументах передавать, громоздкость снизится.

Dimques commented on cs/Markdown/Parser/TokenParser.cs:
```diff
@@ -0,0 +1,316 @@
+        return tokenList[index].Type == type;
+    }
+
+    private static bool IsAtStartOfLine(List<MdToken> tokens, int position)
```
Переименуй `position` ---> `index`. `Position` - это состояние твоего парсера, до куда он допарсил, не перегружай это слово без нужды.

Dimques commented on cs/MarkdownTests/Lexer/LexerTests.cs:
```diff
@@ -0,0 +1,118 @@
+    [TestCaseSource(nameof(LexerTokenizeTestCases))]
+    public void Tokenize_ShouldTokenize_Correctly(string input, List<MdToken> expectedOutput)
+    {
+        var actualInput = MdLexer.Tokenize(input);
```
И где проверка?

Dimques commented on cs/MarkdownTests/MarkdownPerformanceTests.cs:
```diff
@@ -0,0 +1,48 @@
+
+    private static string GenerateRandomMarkdown(int len)
+    {
+        var elements = Enumerable.Range(32, 96).Select(c => ((char)c).ToString()).ToList();
```
Какие-то магические числа `32` и `96`.

Dimques commented on cs/MarkdownTests/MarkdownPerformanceTests.cs:
```diff
@@ -0,0 +1,48 @@
+        {
+            var markdown = GenerateRandomMarkdown(length);
+            sw.Start();
+            GC.Collect();
```
Нуууу такое, конечно. А между сборками мусора будет работать, как попало, да?

Dimques commented on cs/MarkdownTests/MarkdownFuncTests.cs:
```diff
@@ -0,0 +1,80 @@
+    }
+    
+    [Test]
+    [TestCase(@"Escaped \_underscore\_", "Escaped _underscore_")]
```
1. Не работает кейс `@"Escaped _underscore\__"` - я экранировал первый закрывающий прочерк, значит, он должен как текст интерпретироваться и я ожидаю получить `"Escaped <em>underscore_</em>"`.
2. В кейсе `@"\ \"` исчезает первый слэш, как будто он экранирует пробел. Но тогда в кейсе `@"_underscore\  _"`  слово должно оборачиваться в `<em>` вместе с пробелом. Короче, тут либо одно, либо другое неверно - мне кажется, что пробельные символы не должны относиться к символам разметки и подлежать экранированию.
3. В кейсе `@"[text\\](example.com)"` я экранировал слэш внутри значимого текста перед закрывающей квадратной скобкой, ожидаю, что получу `"<a href=\"example.com\">text\\</a>"`, но получаю `"[text\](example.com)"`.
4. Реализация не считает табуляцию пробельным символом, в кейсе `"_\t_"` пытается обернуть её в `<em>`.
5. Почему-то звёздочка экранируется.
6. Почему-то экранируется двоеточие (возможно и другие знаки препинания тоже).
7. Строка из одного-единственного грида точно правильно обрабатывается?

maximka200 committed:
- пофиксил основные замечания
- дописал тестов на ранее неработающие сценарии
- разбил TokenParser.cs на регионы для читаемости

maximka200 committed:
- переписал структуру генерации Html
- переписал тесты под нее

maximka200 committed:
- переписал на Try метод

maximka200 committed:
- поправил тесты

Dimques commented on cs/MarkdownTests/MarkdownFuncTests.cs:
```diff
@@ -0,0 +1,89 @@
+    [TestCase("#  H1\n##  H2\n###  H3", "<h1>H1</h1><br/><h2>H2</h2><br/><h3>H3</h3>")]
+    [TestCase("# _text_ __text__", "<h1><em>text</em> <strong>text</strong></h1>")]
+    [TestCase("  ## h1", "  ## h1")]
+    [TestCase("########### h1", "########### h1")]
```
Здесь лучше было использовать ровно семь гридов, чтобы показать границу числа гридов, начиная с которой другое поведение будет, т.е. подчеркнуть edge-case.

Dimques commented on cs/MarkdownTests/MarkdownFuncTests.cs:
```diff
@@ -0,0 +1,89 @@
+    [TestCase("# _text_ __text__", "<h1><em>text</em> <strong>text</strong></h1>")]
+    [TestCase("  ## h1", "  ## h1")]
+    [TestCase("########### h1", "########### h1")]
+    [TestCase("#", "<h1></h1>")] //  в оригинальном Md также
```
В общем случае это не аргумент - "все совпадения случайны" :)

Dimques commented on cs/MarkdownTests/MarkdownFuncTests.cs:
```diff
@@ -0,0 +1,89 @@
+    [Test]
+    [TestCase("Text with _underscores_ inside", "Text with <em>underscores</em> inside")]
+    [TestCase("__Bold__ _Italic_ __Mixed__", "<strong>Bold</strong> <em>Italic</em> <strong>Mixed</strong>")]
+    [TestCase("root1_2_3", "root1_2_3")]
```
Не работает кейс
<img width="1369" height="54" alt="image" src="https://github.com/user-attachments/assets/b67747b6-f870-4a60-89c3-020c9bc74daa" />

Dimques commented on cs/MarkdownTests/MarkdownFuncTests.cs:
```diff
@@ -0,0 +1,89 @@
+    }
+    
+    [Test]
+    [TestCase(@"Escaped \_underscore\_", "Escaped _underscore_")]
```
Существует, скажем так, правило хорошего тона - ишуи резолвит тот, кто их завёл.

maximka200 replied:
> Существует, скажем так, правило хорошего тона - ишуи резолвит тот, кто их завёл.

понял, просто удобно, проще понимать что сделал/что нет

Dimques commented on cs/MarkdownTests/MarkdownFuncTests.cs:
```diff
@@ -0,0 +1,89 @@
+    
+    [Test]
+    [TestCase("TEXT!???", "TEXT!???")]
+    [TestCase("_Test_\n## Header ##\n__word__ _word_",
```
`\n` - эх, проворонил такой повод подушнить :( На будущее - используй лучше `Environment.NewLine`, в разных системах оно разное значение имеет.

Dimques commented on cs/MarkdownTests/MarkdownPerformanceTests.cs:
```diff
@@ -0,0 +1,58 @@
+            var markdown = GenerateRandomMarkdown(length);
+            sw.Start();
+            Render(markdown);
+            GC.Collect();
```
Ну уж если хватило сил все замечания исправить, мог бы и эту ерунду поправить :)

Dimques commented on cs/MarkdownTests/MarkdownFuncTests.cs:
```diff
@@ -0,0 +1,89 @@
+    [TestCase(@"\\[text](example.com)", """\<a href="example.com">text</a>""")]
+    [TestCase(@"Escaped _underscore\__", "Escaped <em>underscore_</em>")]
+    [TestCase(@"[text\\](example.com)", """<a href="example.com">text\</a>""")]
+    [TestCase("_underscore _", "_underscore _")]
```
Ну кажется на tab надо ещё внимательно посмотреть
<img width="1561" height="215" alt="image" src="https://github.com/user-attachments/assets/0a0221f2-ab7a-41e6-97b9-de2683fc7c1f" />
